### PR TITLE
[feature-wip](mv lifecycle) Support drop partition from index.

### LINKED ISSUE #18505: [Enhancement] separate life cycle of base table and its materialized views
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Doris stores materialized views as tables, however when a table is removed, its materialized views are also removed. Sometimes, users want to keep materialized views when a table is removed, so we should separate life cycle of a table and  its materialized views.

Currently, doris drops partition via `ALTER TABLE [database.]table DROP PARTITION [IF EXISTS] partition_name [FORCE]` which drops tablets of base table, materialized views and rollups. 

### Solution

- [ ] Syntax support `ALTER TABLE [database.]table DROP PARTITION [IF EXISTS] partition_name [FORCE] FROM INDEX [index_name]` @nanfeng1999
- [ ] add index info in DropPartitionInfo @reswqa 
- [ ] Handle index in Env.onErasePartition @reswqa 
- [ ] Pass index info to Env.onErasePartition and DropPartitionInfo. @reswqa 
- [ ] Return user friendly error message for query on dropped partitions which can not be served by materialized index. @zyrate 

### Need Knowledge

- Code creating a materialized view.
- Code using a materialized view when query.
- Data structure related to a materialized view.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by reswqa**
# Proposed changes

Issue Number: partially close #18505 

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

1. Support deleting only partial indices of partition.
2. When the last index of a partition is deleted, delete this partition.

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/common/ErrorCode.java [hunk 1]
```diff
@@ -1192,7 +1192,9 @@ public enum ErrorCode {
      + " table supports time travel in current version"),
 
     ERR_NONSSL_HANDSHAKE_RESPONSE(5091, new byte[] {'4', '2', '0', '0'},
-            "SSL mode on but received non-ssl handshake response from client.");
+            "SSL mode on but received non-ssl handshake response from client."),
+
+    ERR_DROP_PARTITION_FROM_INDEX_NON_EXISTENT(5092, new byte[]{'H', 'Y', '0', '0', '0'}, "Index '%s' is not existent for partition %s.");
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/common/ErrorCode.java:1197]: INDEX NOT EXIST is a better name.

Comment (reswqa) [reply fe/fe-core/src/main/java/org/apache/doris/common/ErrorCode.java]: Renamed to `ERR_NO_SUCH_INDEX` as we have a similar errorCode called `ERR_NO_SUCH_PARTITION`.

**Comment by reswqa** [comment]
Thanks @dataroaring for the review. 

I have updated this pull request in some new commits for the convenience of review, we can squashed them when merging code.  Regression test is also modified to cover this changes. Please take a look in your free time, thanks~

**Comment by morrySnow** [comment]
could we do select rollup/mv correctly after we do drop index partition?

**Comment by reswqa** [comment]
> could we do select rollup/mv correctly after we do drop index partition?

Sorry, I don't fully understand what you mean. Do you mean that after deleting baseIndex, can we still perform queries based on mv?

**Comment by dataroaring** [comment]
run buildall

**Comment by morrySnow** [comment]
> > could we do select rollup/mv correctly after we do drop index partition?
> 
> Sorry, I don't fully understand what you mean. Do you mean that after deleting baseIndex, can we still perform queries based on mv?

yes, i don't see any update about mv/rollup selection, so what' will happen when we have a table with mv and some partitions of mv have been dropped?

**Comment by reswqa** [comment]
@morrySnow Good question. In this PR's implementation, only after all indexes of a partition are deleted will the partition be deleted (The effect is the same as the `drop partition` statement). As for only partial index were deleted:
- If we only drop `base Index` and keep `mv index` for a specific partition, then queries that can hit mv can work correctly. This should be covered by the modified regression test in this PR as following:
```
     // drop base index.
     sql""" ALTER TABLE ${testTable} DROP PARTITION p1 FROM INDEX ${testTable} """
     explain {
       sql(" SELECT k1, k2+k3 FROM ${testTable} PARTITION(p1); ")
       contains "(${testMv})"
     }
     qt_select """ SELECT k1, k2+k3 FROM ${testTable} PARTITION(p1) """
```
- If we only drop `mv index` for a specific partition. After checking the code, I found queries that can hit mv will indeed encounter some problems, such as NPE. Not sure if it requires significant changes to be compatible with this situation as I am not familiar with the selection algorithm of materialized view. If that's the case, I somehow tend to temporarily not support deleting non-based index through the `Drop partition from index` statement. One main reason is that our initial purpose in doing this work was to separate the lifecycle of mv and base table, so that the query can still be performed based on mv. WDYT?

If I missed something, please feel free to point me out 🙇.  Also cc @dataroaring for the more advanced insights.

**Pull request closed without merging**